design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+1. The upgrade controller begins a reconciliation to bring the clusters actual version value in agreement with the desired version from the cluster CRD.
+1. **Operator:** The operator pod itself is upgraded first since it is the host of the upgrade controller.
+If there is any new upgrade logic or any migration needed, the new version of the upgrade controller would know how to perform it, so it needs to be updated first.
+The controller uses the Kubernetes API to update the image field of the operator's pod template, which will then begin the process of the deployment that manages the operator pod to terminate the pod and start a new one running the new version in its place.
+    1. When the operator pod comes back up, it needs to remember that it is in an active upgrade process that needs to be resumed.
+    1. The upgrade controller will verify that the operator pod is on the new version and is in a healthy running state.
+1. **API:** The API pod will be upgraded next by updating the `image` field of the pod template spec.
+The `ROOK_VERSION_TAG` env var will also be updated within the template spec.
+The deployment managing the API pod will then terminate the old pod and then start a pod on the new version to replace it.
+    1. The controller will verify that the new API pod is in the `Running` state with the new version and that basic routes are accessible such as `/status`.
+1. **Mons:** The monitor pods will be upgraded in a rolling fashion next.  **For each** monitor, the following actions will be performed by the upgrade controller:
+    1. The `image` field of the pod template spec will be updated to the new version number.
+    Then the pod will be terminated, allowing the replica set that is managing it to bring up a new pod on the new version to replace it.
+    1.  The controller will verify that the new pod is on the new version, in the `Running` state, and that the monitor returns to `in quorum` and has a Ceph status of `OK`.
+    The cluster health will be verified as a whole before moving to the next monitor.
+1. **OSDs:** The OSD pods will be upgraded in a rolling fashion after the monitors.  **For each** OSD, the following actions will take place:
+    1. The `image` field of the pod template spec will be updated to the new version number.
+    1. The lifecycle management of OSDs can be done either as a whole by a single daemon set or individually by a replica set per OSD.
+    In either case, each individual OSD pod will be terminated so that its managing controller will respawn a new pod on the new version in its place.
+    1. The controller will verify that each OSD is running the new version and that they return to the `UP` and `IN` statuses.
+    Placement group health will also be verified to ensure all PGs return to the `active+clean` status before moving on.
+1. **Ceph Manager:** The Ceph manager pod will be upgraded next by updating the `image` field on the pod template spec.
```

Comment:
the managers are tied fairly closely with the mons and osds. You may need to upgrade them before the mons, and likely before the osds.

Reply:
Good point.  I consulted the [Ceph Luminous upgrade guide](http://ceph.com/releases/v12-1-4-luminous-rc-released/#upgrading) and it recommends doing the mgrs after the mons and before the OSDs.  I'll update my sequence here.